rules/opentitan_test.bzl

```diff
@@ -452,3 +454,138 @@ def opentitan_functest(
+    )
+
+    # Write a GDB script to disk and load it with GDB's `--command` argument.
+    # This enables us to separate lines with whitespace, whereas if we piped the
+    # string into GDB's stdin, each newline would cause it to repeat the
+    # previous command.
+    gdb_script = """target extended-remote :3333
+set pagination off
+
+echo :::: Send OpenOCD the 'reset halt' command.\\n
+monitor reset halt
+
+echo :::: Set RAM as RWX by executing code between _start and test_rom_main()\\n
+file test_rom.elf
+break rom_test_main
+continue
+delete breakpoints
+
+echo :::: Load the SRAM program onto the device and check integrity.\\n
+file sram_program.elf
+load sram_program.elf
+compare-sections
+
+echo :::: Call sram_main().\\n
+print sram_main()
+
+echo :::: Done.\\n
+    """
+
+    test_script_file = ctx.actions.declare_file("{}.sh".format(ctx.label.name))
+    gdb_script_file = ctx.actions.declare_file("{}.gdb".format(ctx.label.name))
+
+    ctx.actions.write(output = test_script_file, content = test_script)
+    ctx.actions.write(output = gdb_script_file, content = gdb_script)
+
+    return [
+        DefaultInfo(
+            executable = test_script_file,
+            runfiles =
+                ctx.runfiles(
+                    # Relative paths provided by `File.path` seem to not work
+                    # for generated files. Symlinking the runtime files wherever
+                    # Bazel will run `test_script_file` sidesteps the issue.
+                    symlinks = {
+                        "gdb_script": gdb_script_file,
+                        "opentitantool": ctx.file.opentitantool,
+                        "sram_program.elf": ctx.file.sram_program_elf,
+                        "test_rom.bit": ctx.file.test_rom_bitstream,
+                        "test_rom.elf": ctx.file.test_rom_elf,
+                    },
+                    files = [
+                        gdb_script_file,
+                        ctx.file.sram_program_elf,
+                        ctx.file.test_rom_bitstream,
+                        ctx.file.test_rom_elf,
+                        ctx.file.opentitantool,
+                        ctx.file.openocd_earlgrey_config,
+                    ],
+                ),
+        ),
+    ]
+
+# Orchestrate opentitantool, OpenOCD, and GDB to load the given program into
+# SRAM and execute it in-place. This rule assumes that a CW310 FPGA and an
+# ARM-USB-TINY-H JTAG debugger are attached to the host.
+opentitan_sram_fpga_cw310_test = rv_rule(
+    implementation = _opentitan_sram_fpga_cw310_test,
+    attrs = {
+        "sram_program_elf": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+        ),
+        "exit_success_pattern": attr.string(
+            default = "sram_program\\.c:47\\] PC: 0x1000011c, SRAM: \\[0x10000000, 0x10020000\\)",
+        ),
+        "test_rom_bitstream": attr.label(
+            default = "//hw/bitstream:test_rom",
```

Comment:
@alphan For realism, shouldn't we be using the ~~Mask~~ ROM?

Reply:
Yes

Reply:
But you may want to do this as a follow up since this would affect the script and also ePMP configuration.